wsmaster/che-core-api-machine/src/main/java/org/eclipse/che/api/machine/server/MachineManager.java

```diff
@@ -326,37 +343,42 @@ private MachineImpl createMachine(MachineConfigImpl machineConfig,
                                                     null);
 
         createMachineLogsDir(machineId);
-        final LineConsumer machineLogger = getMachineLogger(machineId, getMachineChannels(machine.getConfig().getName(),
-                                                                                          machine.getWorkspaceId(),
-                                                                                          machine.getEnvName())
-                .getOutput());
+        final LineConsumer machineLogger = getMachineLogger(machineId, outputConsumer);
 
         try {
+            machineRegistry.addMachine(machine);
             try {
-                machineRegistry.addMachine(machine);
                 instanceCreator.createInstance(instanceProvider, machine, machineLogger);
             } catch (MachineException ex) {
                 if (snapshot == null) {
                     throw ex;
                 }
                 if (ex.getCause() instanceof SourceNotFoundException) {
-                    final LineConsumer logger = getMachineLogger(machineId,
-                                                                 getMachineChannels(machine.getConfig().getName(),
-                                                                                    machine.getWorkspaceId(),
-                                                                                    machine.getEnvName()).getOutput());
                     LOG.error("Image of snapshot for machine " + machineConfig.getName() + " not found. " +
                               "Machine will be created from origin source");
                     machine.getConfig().setSource(sourceCopy);
+                    try {
+                        machineRegistry.remove(machineId);
+                    } catch (NotFoundException ignored) {
+                        // machine is already removed, should never happen
+                    }
                     machineRegistry.addMachine(machine);
-                    instanceCreator.createInstance(instanceProvider, machine, logger);
+                    instanceCreator.createInstance(instanceProvider, machine, outputConsumer);
                 }
             }
             return machine;
-        } catch (ConflictException e) {
+        } catch (ApiException e) {
             try {
                 machineLogger.close();
-            } catch (IOException ignored) {
+            } catch (IOException e1) {
```

Comment:
My preference:
when you have nested exceptions it is always good to use significant names, as if you need to reuse nested exception it may be confusing to deal with _e_, _e1_, _e2_ etc. So i would rename `e -> apiEx`, `e1 -> ioEx`.